### PR TITLE
opt dyn_slice/dyn_pad codegen using shape-constraint

### PR DESCRIPTION
**Pull request by wyzero**

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.h [hunk 1]
```diff
@@ -87,6 +87,102 @@ inline bool operator!=(const SymbolicDimProduct& lhs,
...
+
+  // Returns  the attribute name used when saving.
+  // An exam ple is
+  //   "mhlo .real_dynamic_slice"(...) { kDiscSliceOpStaticKnownInfo = {...}}
+  static StringRef getAttrName() { return "kDiscSliceOpStaticKnownInfo"; }
+
+  // Saves the updates attribute to the `op` this helper holds.
+  LogicalResult save();
+
+  enum {
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.h:103]: nit: a named enum class? `kUnknown` looks pretty easy to get confilct..

File: tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.cc [hunk 2]
```diff
@@ -888,5 +889,208 @@ template <typename Combiner>
...
+  for (int i = 0; i < ty.getRank(); ++i) {
+    mergeStartIndex(i, startAttr.getValues<int64_t>()[i]);
+    mergeLimitIndex(i, limitAttr.getValues<int64_t>()[i]);
+    mergeStride(i, strideAttr.getValues<int64_t>()[i]);
+  }
+}
+
+bool SliceOpShapeHelper::isFullySlicedAxis(int axis) {
+  assert(axis < static_cast<int>(startIndices.size()));
+  return startIndices[axis] == 0 && limitIndices[axis] == 1 &&
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.cc:920]: `limitIndices[axis] == 1` -> `strides[axis] == 1` ?

File: tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.cc [hunk 3]
```diff
@@ -888,5 +889,208 @@ template <typename Combiner>
...
+}
+
+LogicalResult SliceOpShapeHelper::mergeStartIndex(int axis, int64_t value) {
+  assert(axis < static_cast<int>(startIndices.size()));
+
+  if (startIndices[axis] != value && value != kUnknown &&
+      startIndices[axis] != kUnknown)
+    return failure();
+
+  if (startIndices[axis] == kUnknown) startIndices[axis] = value;
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.cc:949]: if `value` is `kUnknown` and `startIndices[axis]` is not, it succeeded though no merge happend. it this expected?

Comment (qiuxiafei) [reply tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.cc:949]: fine... it's initialized to `kUnknown`.

File: tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc [hunk 4]
```diff
@@ -561,15 +591,35 @@ Value elementalLower<lmhlo::DynamicPadOp>(OpBuilder* b, Location loc,
...
+    }
+
     SmallVector<Value, 4> dim_const;
     dim_const.push_back(b->create<arith::ConstantIndexOp>(loc, dim));
-    Value edge_padding_low =
-        createMaySpecificLoad(*b, loc, op.getOperation(),
-                              edge_padding_low_memref, dim_const, lower_config);
+    Value edge_padding_low;
+    if (useShapeConstraintIR() &&
+        helper->edgePaddingLows[dim] != PadOpShapeHelper::kUnknown) {
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc:603]: nit: would prefer a method like `helper->isUnknownEdgePaddingLows()`. :P

**Review by qiuxiafei — APPROVED** [review]
some minor comments, generally LGTM.

**Review by qiuxiafei — APPROVED** [review]
some minor comments, generally LGTM.

File: tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.h (hunk 1 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.h]: Done

File: tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.cc (hunk 2 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.cc]: Done, thanks!

File: tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc (hunk 4 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/lhlo_elemental_utils.cc]: Done!

**Pull request merged**